### ISSUE TITLE
Bug Report: with_items and selectattr filter using equalto test.

### ISSUE THREAD

**Issue by jgornick** [ansible/ansible#8836]
##### Issue Type:

Bug Report
##### Ansible Version:

1.7.1
##### Environment:

OS X 10.9.4
##### Summary:

When trying to use the `selectattr` filter, using the `equalto` test, in conjunction with `with_items`, it doesn't seem to properly evaluate the selectattr functionality as described in documentation. 
##### Steps To Reproduce:

Run the following playbook: https://gist.github.com/jgornick/0586deadad6bb2693f03
##### Expected Results:

I would of expected to see one debug message for the "user1" user in the list I filtered.
##### Actual Results:

Output when running the playbook: 

```
TASK: [debug var=item
] ******************************************************* 
ok: [127.0.0.1] => (item=users | selectattr("username", "equalto", "user1") | list
) => {
    "item": "users | selectattr(\"username\", \"equalto\", \"user1\") | list\n"
}
```

**Comment by jgornick** [comment]
After some further tests, it doesn't seem like `selectattr` while using the `equalto` test works anywhere. Is this an issue with the bundled version of jinja2?

**Comment by jimi-c** [comment]
I actually ran into this today. The 'equalto' test is only available in the 2.8+ version of jinja2. Because of this, I started working on adding custom tests to core (since we have custom filters already), to add a stop-gap solution. I'll update this ticket if/when we merge that in.

**Comment by jgornick** [comment]
@jimi-c Thank you!

**Comment by adamhadani** [comment]
Any updates on this? Would love to see selectattr with equalto or an equivalent , highly useful!

**Comment by aiwilliams** [comment]
:+1:

**Comment by tcurdt** [comment]
I've also run into this. That said - where is the documentation that says it should be "equalto"? what other test are available?

**Comment by tcurdt** [comment]
more information:

http://stackoverflow.com/questions/20529234/how-to-select-reduce-a-list-of-dictionaries-in-flask-jinja

**Comment by tcurdt** [comment]
so maybe

```
app.jinja_env.tests['equalto'] = lambda value, other : value == other
```

works as a work around?

**Comment by tcurdt** [comment]
As a work around I wanted create a custom filter but as it turns out even an invalid filter `{{user|blub}}` seems to turn the item into a string. Shouldn't that throw an error?

**Comment by tcurdt** [comment]
As a work around I created a `filter_plugins` directory in my playbook's folder. In there I have a `core.py` (name should not matter) that looks like this:

```
def filter_list(list, value):
    return filter(lambda t: t['type'] == value, list)

class FilterModule(object):
    def filters(self):
        return {
            'type': filter_list
        }
```

What a PITA figuring this out.

**Comment by micw** [comment]
@tcurdt Could you please explain, how to (exactly) use that workaround to solve the problem above? can put this file also somewhere to my roles?

Edit: found it out by myself. I changed the code:

def filter_list(list, key, value):
    return filter(lambda t: t[key] == value, list)

class FilterModule(object):
    def filters(self):
        return {
            'byattr': filter_list
        }

It can be used so:
with_items: users | byattr("username", "user1")

**Comment by quixoten** [comment]
re: #9645, which allows custom jinja2 tests to be defined as test plugins.

**Comment by zkf** [comment]
@tcurdt, @micw: thanks a lot for the workaround :thumbsup:

**Comment by jimi-c** [comment]
Going to go ahead and close this in favor of your work in #9645, even though we'll have to do a bit of massaging to get that into the v2 code base (which I'll take care of).

Thanks!

**Comment by RusDavies** [comment]
Be careful with the filenames of custom filter plugins under `./filter_plugins/` in Ansible 2.0.0.  

Unlike in 1.9.x, if you give the filter plugin a filename that already exists under `ansible/lib/ansible/plugins/filter/`, (and possibly elsewhere), such as `core.py` suggested earlier in this thread, then strange things happen. See Ansible issue [127238](https://github.com/ansible/ansible/issues/12738).

In short, be sure to give your filter plugins unique file names.

**Comment by micw** [comment]
@RusDavies thank you for that hint. Are filter_plugins automatically loaded when they are named different than core.py?

**Comment by RusDavies** [comment]
@micw,  yeap.

**Comment by SergeyBear** [comment]
More natural way for Ansible is to create /etc/ansible/test_plugins/custom.py with content

```
from __future__ import (absolute_import, division, print_function)
__metaclass__ = type

from ansible import errors

def equalto(value, other):
    return bool(value == other)

class TestModule(object):
    ''' Ansible file jinja2 tests '''

    def tests(self):
        return {
            'equalto' : equalto,
        }
```
